presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,224 @@
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            List<String> partitionsInFileSystem = listDirectory(fileSystem, fileSystem.getFileStatus(defaultLocation), table.getPartitionColumns(), table.getPartitionColumns().size()).stream()
+                    .map(fileStatus -> defaultLocation.toUri().relativize(fileStatus.getPath().toUri()).getPath())
+                    .collect(toImmutableList());
+
+            // partitions in metastore but not in file system
+            Set<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            // partitions in file system but not in metastore
+            Set<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+
+            if (syncMode == SyncMode.ADD || syncMode == SyncMode.FULL) {
+                addPartitions(metastore, session, table, defaultLocation, partitionsToAdd);
+            }
+            if (syncMode == SyncMode.DROP || syncMode == SyncMode.FULL) {
+                dropPartitions(metastore, session, table, partitionsToDrop);
+            }
+            metastore.commit();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<FileStatus> listDirectory(FileSystem fileSystem, FileStatus current, List<Column> partitionColumns, int depth)
+    {
+        if (depth == 0) {
+            return ImmutableList.of(current);
+        }
+
+        try {
+            return Stream.of(fileSystem.listStatus(current.getPath()))
+                    .filter(fileStatus -> validatePath(fileSystem, fileStatus, partitionColumns.get(partitionColumns.size() - depth)))
+                    .flatMap(directory -> listDirectory(fileSystem, directory, partitionColumns, depth - 1).stream())
+                    .collect(toImmutableList());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean validatePath(FileSystem fileSystem, FileStatus file, Column column)
+    {
+        try {
+            Path path = file.getPath();
+            String prefix = column.getName() + '=';
+            return fileSystem.isDirectory(path) && path.getName().startsWith(prefix);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    // calculate relative complement of set b with respect to set a
+    private Set<String> difference(List<String> a, List<String> b)
+    {
+        return Sets.difference(newHashSet(a), newHashSet(b));
```

Comment:
`newHashSet()` and related are considered deprecated on Java 7+ since the diamond operator does the same thing. Use `new HashSet<>(x)` instead.